### ISSUE TITLE
Handshake breaks backward compatibility

### ISSUE THREAD

**Issue by bowenwang1996** [nearprotocol/nearcore#3283]
This line https://github.com/nearprotocol/nearcore/blob/3652a0901a280b9b5da3ccde920c1fd7557c7732/chain/network/src/peer.rs#L683 breaks backward compatibility because nodes with newer version will not be able to connect to nodes with even one version older.

**Comment by SkidanovAlex** [comment]
When we upgrade to HandhakeV2, we should make old Handshake message to include the oldest supported version, not the current version. This will allow connecting to the old peers.
CC @mfornet

**Comment by mfornet** [comment]
@bowenwang1996 The workflow is as following: 

There are two nodes,  A with version=4 and oldest_supported_version=3 and B with version=3.
- A initiates a handshake to B and let B knows it is using version 4.

- Since version 4 is newer and unknown to B, then B respond to A with:

```rust
PeerMessage::HandshakeFailure(_, 
    HandshakeFailureReason::ProtocolVersionMismatch {
        version: 3,
        oldest_supported_version: 3
 })
```

- A learns that B is using version=3, and starts serializing and deserializing all messages to B with version=3. Even handshake.

Basically, A instead of sending its current protocol version in the handshake, it will use the highest protocol that both A and B speaks. Notice that to update the `PROTOCOL_VERSION` for any running peer it will require the client to be restarted.

**Comment by bowenwang1996** [comment]
So we agreed to address this in the following way:
* Merge #3157. This will allow nodes to recognize older versions in the next release. Since both betanet and testnet already contains #3016, this will work fine.
* For mainnet, we need to do a two-step upgrade. First we need to upgrade nodes to a version that contains #3016. After that is done, we do a second upgrade to a newer version that contains #3157.

**Comment by bowenwang1996** [comment]
This is mostly resolved in #3157. The rest of the work is laid out in https://github.com/nearprotocol/nearcore/issues/3289
